Source/Core/VideoCommon/RenderBase.cpp

```diff
@@ -406,6 +406,21 @@ bool Renderer::IsHeadless() const
   return !m_surface_handle;
 }
 
+void Renderer::ChangeSurface(void* new_surface_handle)
+{
+  std::lock_guard<std::mutex> lock(m_surface_change_mutex);
+  m_new_surface_handle = new_surface_handle;
+  m_surface_needs_change.Set();
+}
```

Comment:
Is it allowed to still use the old surface after this call completes? Else we need to block till the backend really switched the surfaces. Or at least while the backend is presenting its data.

Reply:
I didn't want to block the UI thread here as this could lead to a deadlock if the video thread is waiting for a message to be processed. But yeah, even with the flag there is still a race. Holding the lock while presenting the surface is probably the safest way around this, but AFAICT to avoid the deadlock, it would have to be with a try_lock, periodically yielding to the message pump.

Reply:
I think the deadlock will be much more unlikely if we only block if the video thread is in the swapping code. Can the SwapImpl() call block at all?

Reply:
Hmm, upon thinking about it further, if what I was saying is the case, then single-threaded applications which have the message loop and rendering on the same thread would break. So I think making SwapImpl and ChangeSurface mutually exclusive should be sufficient.